### ISSUE TITLE
pygit2 doesn't see all changed files

### ISSUE THREAD

**Issue by bochecha** [libgit2/pygit2#535]
Using git:

```
$ git diff --name-only 
hm/__init__.py
hm/cli.py
```

But with pygit2:

```
>>> import pygit2
>>> repo = pygit2.Repository('.')
>>> index = repo.index
>>> index.read()
>>> list(index.diff_to_workdir())
[<_pygit2.Patch object at 0x7fc809008a30>]
>>> list(index.diff_to_workdir())[0].old_file_path
'hm/__init__.py'
>>> list(index.diff_to_workdir())[0].new_file_path
'hm/__init__.py'
```

As you can see, pygit2 is completely missing the changes that happened in `hm/cli.py`.

Or am I doing this wrong? :confused:

**Comment by carlosmn** [comment]
You're only inspecting the first entry from the list of differences, so it's right that you're getting information about a single file. If you look further into the list, you'll see the other file.

**Comment by bochecha** [comment]
Well, sure, but the list only has one item.

**Comment by carlosmn** [comment]
Ah yes, I missed that line among the file listing. The prime suspect in that case would be crlf filtering you might have on (or if you have other filters turned on).

**Comment by bochecha** [comment]
How do I check that?

**Comment by bochecha** [comment]
@carlosmn did you mean something like this? https://libgit2.github.com/libgit2/#HEAD/type/git_filter

I don't think I'm doing anything with regards to these filters in my code, at least not explicitly.

I didn't even find any API that looks like that in pygit2, so I'm really not sure how I could have set such filters in my code. :smile: 

Nevertheless, is there a way to inspect what's going on, whether there are some filters applied, and how to fix things?

**Comment by carlosmn** [comment]
`git_filter` is for creating your own in-code filters to be called by libgit2. They would not affect git or its idea of changes. Filter for git are things like crlf or the clean and smudge filters you can have for stuff like compressing files or encrypting them which libgit2 would not take into account. There's also textconv but IIRC that does not change the idea of change, just the textual output.

What *are* the differences in that file that doesn't get recognised? There are times when some versions of git will show an empty diff for a particular file due to implementation details, which libgit2 would not do.

**Comment by carlosmn** [comment]
There is another possibility, as some versions of libgit2 don't fully implement racy-git checking, if `hm/cli.py` has the same mtime timestamp as the index (`.git/index`) it's possible that we're failing to notice that there are differences.

**Comment by bochecha** [comment]
> What are the differences in that file that doesn't get recognised?

Actual code changes: lines added, lines removed.

> There is another possibility, as some versions of libgit2 don't fully implement racy-git checking, if hm/cli.py has the same mtime timestamp as the index (.git/index) it's possible that we're failing to notice that there are differences.

```
>>> os.path.getmtime('.git/index')
1434262969.9596527
>>> os.path.getmtime('hm/cli.py')
1434360495.7066977
```

So `hm/cli.py` has a newer mtime than the index.

By the way, about versions:

```
>>> pygit2.LIBGIT2_VERSION
'0.22.2'
>>> pygit2.__version__
'0.22.0'
```

I should probably have said that right from the start. :smile:

**Comment by jdavid** [comment]
Cannot reproduce, maybe this has been fixed since 0.22
Closing, sorry for not handling this at the time.
